src/java/org/apache/cassandra/service/accord/CommandsForRanges.java

```diff
@@ -160,33 +158,279 @@ private <P1, T> T mapReduce(@Nonnull Timestamp testTimestamp, @Nullable TxnId te
+                }
+            }
+        }
+
+        public void forEachInCache(Consumer<Summary> forEach, AccordCommandStore.Caches caches)
+        {
+            for (TxnId txnId : manager.cachedRangeTxns)
+            {
+                AccordCacheEntry<TxnId, Command> state = caches.commands().getUnsafe(txnId);
+                Summary summary = from(state);
+                if (summary != null)
+                    forEach.accept(summary);
+            }
+        }
+
+        public Summary load(TxnId txnId)
+        {
+            if (findAsDep == null)
+            {
+                SavedCommand.MinimalCommand cmd = manager.commandStore.loadMinimal(txnId);
+                if (cmd != null)
+                    return from(cmd);
+            }
+            else
+            {
+                Command cmd = manager.commandStore.loadCommand(txnId);
+                if (cmd != null)
+                    return from(cmd);
+            }
+
+            Ranges ranges = manager.transitive.get(txnId);
+            if (ranges == null)
+                return null;
+
+            ranges = ranges.intersecting(searchKeysOrRanges);
+            if (ranges.isEmpty())
+                return null;
+
+            return new Summary(txnId, txnId, SaveStatus.NotDefined, ranges, null, false);
+        }
+
+        public Summary from(AccordCacheEntry<TxnId, Command> state)
+        {
+            if (state.key().domain() != Routable.Domain.Range)
+                return null;
+
+            switch (state.status())
+            {
+                default: throw new AssertionError("Unhandled status: " + state.status());
+                case LOADING:
+                case WAITING_TO_LOAD:
+                case UNINITIALIZED:
+                    return null;
+
+                case LOADED:
+                case MODIFIED:
+                case SAVING:
+                case FAILED_TO_SAVE:
+            }
+
+            TxnId txnId = state.key();
+            if (!txnId.isVisible() || txnId.compareTo(minTxnId) < 0 || txnId.compareTo(maxTxnId) >= 0)
+                return null;
+
+            Command command = state.getExclusive();
+            if (command == null)
+                return null;
+            return from(command);
+        }
+
+        public Summary from(Command cmd)
+        {
+            return from(cmd.txnId(), cmd.executeAt(), cmd.saveStatus(), cmd.participants(), cmd.partialDeps());
+        }
+
+        public Summary from(SavedCommand.MinimalCommand cmd)
+        {
+            Invariants.checkState(findAsDep == null);
+            return from(cmd.txnId, cmd.executeAt, cmd.saveStatus, cmd.participants, null);
+        }
+
+        private Summary from(TxnId txnId, Timestamp executeAt, SaveStatus saveStatus, StoreParticipants participants, @Nullable PartialDeps partialDeps)
+        {
+            if (participants == null)
+                return null;
+
+            Ranges keysOrRanges = participants.touches().toRanges();
```

Comment:
previously I have seen `touches` have more ranges than `owns`, so wondering why we do this rather than `owns`?   previously we would load then `slice` to what is logically `owns`, so wondering why this change?  we added slice to CFR as we were including ranges outside of owns and you previously said that was a bug, so feels like we are allowing that to come back?

Reply:
because we rely on quorum intersections across topology changes. "Touches" means we have processed a phase (e.g. a PreAccept, Accept, Commit etc) that intersected with ranges we own _at some point_, and we did this because some future epoch broadcast backwards to us, to ensure commands we process intersect these later command's quorums. This means for both recovery and for dependency calculation, anyone asking for information about a command intersecting those ranges needs to encounter this command. 

Ownership means we're responsible _for that command_
Touches means we're responsible for some _interactions_ with that command

Touches is always bigger than owns, but still represents ranges we're responsible for.